lib/auth/register.go

```diff
@@ -207,173 +217,236 @@ func registerThroughProxy(token string, params RegisterParams) (*Identity, error
-	// but if it's not available connect without validating the Auth Server CA.
-	switch {
-	case len(params.CAPins) != 0:
-		client, err = pinRegisterClient(params)
-	default:
-		client, err = insecureRegisterClient(params)
-	}
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// first, try to read given CA path
+	if cfg.CAPath != "" {
+		rawClusterCAs, err = utils.ReadPath(cfg.CAPath)
+		if err != nil && !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
 	}
-	defer client.Close()
 
-	// Get the SSH and X509 certificates for a node.
-	certs, err := client.RegisterUsingToken(types.RegisterUsingTokenRequest{
-		Token:                token,
-		HostID:               params.ID.HostUUID,
-		NodeName:             params.ID.NodeName,
-		Role:                 params.ID.Role,
-		AdditionalPrincipals: params.AdditionalPrincipals,
-		DNSNames:             params.DNSNames,
-		PublicTLSKey:         params.PublicTLSKey,
-		PublicSSHKey:         params.PublicSSHKey,
-		EC2IdentityDocument:  params.EC2IdentityDocument,
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// If given a proxy address the client will ask the proxy to use TLS
+	// routing to forward the connection to the auth server. Fetch the
+	// cluster name and the auth server's tls certs here to configure the
+	// connection.
+	for _, proxyAddr := range cfg.Addrs {
+		clusterCAResponse, err := webclient.GetClusterCA(ctx, proxyAddr, cfg.Insecure)
+		if err == nil {
+			clusterName = clusterCAResponse.ClusterName
+			rawClusterCAs = append(rawClusterCAs, clusterCAResponse.ClusterCAs...)
+			break
+		}
+		log.WithError(err).Debugf("Failed to retrieve cluster CA certs from proxy "+
+			"endpoint, %q is probably not a proxy", proxyAddr)
 	}
 
-	return ReadIdentityFromKeyPair(params.PrivateKey, certs)
-}
+	tlsConfig := &tls.Config{
+		Time: cfg.Clock.Now,
+	}
 
-// insecureRegisterClient attempts to connects to the Auth Server using the
-// CA on disk. If no CA is found on disk, Teleport will not verify the Auth
-// Server it is connecting to.
-func insecureRegisterClient(params RegisterParams) (*Client, error) {
-	tlsConfig := utils.TLSConfig(params.CipherSuites)
-	tlsConfig.Time = params.Clock.Now
+	// parse CAs certs if we have any
+	if len(rawClusterCAs) > 0 {
+		certs, err := tlsca.ParseCertificatePEMs(rawClusterCAs)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if err := validateCertificates(certs, cfg.Clock); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		tlsConfig.RootCAs = x509.NewCertPool()
+		for _, cert := range certs {
+			tlsConfig.RootCAs.AddCert(cert)
+		}
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
 
-	cert, err := readCA(params.CAPath)
-	if err != nil && !trace.IsNotFound(err) {
+	// create a client with certs from above, if any
+	unauthenticatedClient, err := NewClient(client.Config{
+		Addrs: cfg.Addrs,
+		Credentials: []client.Credentials{
+			client.LoadTLS(tlsConfig),
```

Comment:
To leverage SNI Routing on Nginx cloud side the ServerName should contain encoded cluster name like: 
`6461766964342e636c6f75642e6772617669746174696f6e616c2e696f.teleport.cluster.local` or host obtained from the WebProxy address.

Reply:
Set `ServerName` to the encoded cluster name when connecting through a proxy in the latest commit